cpp/src/arrow/filesystem/path_util.cc

```diff
@@ -287,6 +288,38 @@ bool IsLikelyUri(util::string_view v) {
   return ::arrow::internal::IsValidUriScheme(v.substr(0, pos));
 }
 
+struct Globber::Impl {
+  std::regex pattern_;
+
+  explicit Impl(std::string p) : pattern_(std::regex(std::move(transform(p)))) {}
```

Comment:
Nit: `std::move` is not required here.
```suggestion
  explicit Impl(const std::string& p) : pattern_(std::regex(transform(p))) {}
```